Added ECDH-1PU (Draft 04) algorithm support to JWE

spivachuk opened the pull request:
**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
    - Fixed a bug with a wrong size of the agreed upon key in direct key agreement mode of ECDH-ES and ECDH-1PU.
    - Prepended the shared secret computation in OKPKey class with lazy initialization of private_key field.
- [x] Feature
    - Added ECDH-1PU (Draft 04) algorithm support to JWE implementation.
    - Enhanced unit tests of ECDH-ES algorithm.
- [ ] Code style update
- [ ] Refactor
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No (while `JWEAlgorithm` is subclassed within Authlib only)

---

- [x] You consent that the copyright of your pull request source code belongs to Authlib's author.

ashcherbakov commented on authlib/jose/drafts/_jwe_algorithms.py:
```diff
@@ -0,0 +1,161 @@
+    ALLOWED_KEY_CLS = ECKey
+
+    # https://datatracker.ietf.org/doc/html/draft-madden-jose-ecdh-1pu-04
+    def __init__(self, key_size=None):
```
Can `key_size` be any integer? I think it would be better to define an Enum for supported key sizes.

ashcherbakov commented on authlib/jose/rfc7516/jwe.py:
```diff
@@ -75,18 +77,28 @@ def serialize_compact(self, protected, payload, key):
-        ek = wrapped['ek']
-        if 'header' in wrapped:
-            protected.update(wrapped['header'])
+        if isinstance(alg, ECDH1PUAlgorithm) and alg.key_size is not None:
```
Can we move that logic into specific classes such as ECDH-ES and ECDH-1PU, keeping the JWE code unaware of the details, so that there is no need for `isinstance` calls?
I think we will have to do something like this in any case when we start supporting JSON serialization.

ashcherbakov commented on tests/core/test_jose/test_jwe.py:
```diff
@@ -238,6 +251,135 @@ def test_ecdh_es_raise(self):
         )
 
+    def test_ecdh_1pu_key_agreement_computation(self):
+        # https://tools.ietf.org/html/rfc7518#appendix-C
```
Should it be https://datatracker.ietf.org/doc/html/draft-madden-jose-ecdh-1pu-04#appendix-A?

ashcherbakov commented on tests/core/test_jose/test_jwe.py:
```diff
@@ -238,6 +251,135 @@ def test_ecdh_es_raise(self):
             protected, b'hello', key
         )
 
+    def test_ecdh_1pu_key_agreement_computation(self):
```
Please add a similar test for test vectors from https://datatracker.ietf.org/doc/html/draft-madden-jose-ecdh-1pu-04#appendix-B

ashcherbakov commented on tests/core/test_jose/test_jwe.py:
```diff
@@ -238,6 +251,135 @@ def test_ecdh_es_raise(self):
+            "d": "VEmDZpDXXK8p8N0Cndsxs924q6nS1RXFASRl6BfUqdw"
+        }
+        for alg in ["ECDH-1PU", "ECDH-1PU+A128KW", "ECDH-1PU+A192KW", "ECDH-1PU+A256KW"]:
+            protected = {'alg': alg, 'enc': 'A256GCM'}
```
KW algorithms can work with CBC encryptions algorithms only

ashcherbakov commented on tests/core/test_jose/test_jwe.py:
```diff
@@ -238,6 +251,135 @@ def test_ecdh_es_raise(self):
+            rv = jwe.deserialize_compact(data, bob_key, sender_key=alice_key)
+            self.assertEqual(rv['payload'], b'hello')
+
+    def test_ecdh_1pu_raise(self):
```
I believe we may need to add more tests (and possibly checks in implementation) here:
1) Make sure that ECDH-1PU+AXXXKW algorithms can work with CBC encryption algorithms only (A256CBC-HS512 etc.). Direct key mode can work with any algorithms
2) A test that raises an exception if valid keys are passed (private and public) but of different type (JWK key type field)
3) A test that raises an exception if the key types are valid, the format is valid, but the value of the points (x, y) doesn't belong to the same curve.

spivachuk committed:
Added ECDH-1PU key agreement

spivachuk committed:
Added more tests for ECDH-1PU

- Added more tests for ECDH-1PU.
- Made minor corrections in existing tests of ECDH-ES and ECDH-1PU.
- Made a minor correction in JsonWebEncryption class.

spivachuk committed:
Added more tests for ECDH-1PU

spivachuk committed:
Fixed bug with key size in direct key agreement mode

- Fixed a bug with a wrong size of the agreed upon key in direct key agreement mode of ECDH-ES and ECDH-1PU.
- Removed debug output from RSAAlgorithm.

spivachuk committed:
Corrected OKPKey.exchange_shared_key method

- Prepended the shared secret computation in OKPKey class with lazy initialization of private_key field.
- Added verifications with OKPKeys to test_ecdh_1pu_encryption_fails_if_key_points_are_not_actually_on_same_curve.

spivachuk committed:
Minor changes in unit tests of JWE

spivachuk committed:
Parametrized positive tests of ECDH-ES and ECDH-1PU

- Parametrized positive tests of JWE that use ECDH-ES and ECDH-1PU algorithms.
- Made minor corrections in existing tests of JWE.

spivachuk committed:
Removed setting and verifying skid header field

- Removed setting and verifying skid header field from ECDH-1PU algorithm.

ashcherbakov commented on authlib/jose/__init__.py:
```diff
@@ -15,14 +15,14 @@
 from .rfc7518 import (
     register_jws_rfc7518,
```
We need to extend the README with info about ECDH-1PU

ashcherbakov commented on authlib/jose/errors.py:
```diff
@@ -25,12 +25,21 @@ def __init__(self, result):
+    error = 'inappropriate_encryption_algorithm'
+
+    def __init__(self, enc, alg):
+        description = '{} encryption algorithm is inappropriate for {} algorithm'.format(enc, alg)
```
Should we add details about what enc algorithms are expected for this algorithm in the error message?

ashcherbakov commented on authlib/jose/drafts/_jwe_algorithms.py:
```diff
@@ -0,0 +1,163 @@
+            return raw_data
+        return ECKey.import_key(raw_data)
+
+    def compute_fixed_info(self, headers, bit_size, tag):
```
Should all methods in this class be public? Should we make internal methods private (with `_` prefix)?

ashcherbakov commented on tests/core/test_jose/test_jwe.py:
```diff
@@ -166,78 +169,656 @@ def test_ase_gcm_jwe_invalid_key(self):
+
+        alg = JsonWebEncryption.ALG_REGISTRY['ECDH-1PU']
+
+        alice_static_key = alg.prepare_key(alice_static_key)
```
It would be great to check that Ze, Zs, Z, derived key and other parameters are equal to test vector values from Appendix A.

ashcherbakov commented on tests/core/test_jose/test_jwe.py:
```diff
@@ -166,78 +169,656 @@ def test_ase_gcm_jwe_invalid_key(self):
+        protected_segment = json_b64encode(headers)
+        aad = to_bytes(protected_segment, 'ascii')
+
+        ciphertext, tag = enc.encrypt(payload, aad, iv, cek)
```
It would be great to check that Ze, Zs, Z (and all other parameters that are not checked) are equal to test vector values from Appendix B.

spivachuk committed:
Made minor corrections

- Made some methods protected.
- Refined one exception class.

spivachuk committed:
Add more verifications to ECDH-ES and ECDH-1PU tests

- Add more verifications to ECDH-ES and ECDH-1PU tests.
- Made minor refactoring in ECDHESAlgorithm and ECDH1PUAlgorithm classes.
- Added ECDH-1PU (Draft 04) item to README files.
